instrumentation/opentelemetry_elli/README.md

```diff
@@ -0,0 +1,91 @@
+# opentelemetry_elli
+
+![Common Test](https://github.com/opentelemetry-beam/opentelemetry_elli/workflows/Common%20Test/badge.svg) [![Gitter](https://badges.gitter.im/open-telemetry/opentelemetry-erlang.svg)](https://gitter.im/open-telemetry/opentelemetry-erlang?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
+
+Elli middleware for OpenTelemetry instrumentation.
+
+## Setup and Configuration
+
+``` erlang
+{deps, [opentelemetry_elli]}.
+```
+
+Using the elli_middleware callback place oc_elli_middelware as the first module to be called in the list of handlers:
+
+``` erlang
+[{callback, elli_middleware},
+ {callback_args, [{mods, [{otel_elli_middleware, []},
+                          {<YOUR HANDLER>, []}]}]}]
+```
+
+
+
+OpenTelemetry's [HTTP Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#http-server) for a server details the attributes added to a Span automatically by using this middleware. One such attribute must be set in your `sys.config` under the `opentelemetry_elli` application like:
+
+``` erlang
+{opentelemetry_elli, [{server_name, <<"my-http-server">>}]}.
+```
+
+It is strongly recommended to set this environment variable so the attribute can be included:
+
+> http.server_name has shown great value in practice, as bogus HTTP Host headers occur often in the wild. It is strongly recommended to set http.server_name to allow associating requests with some logical server entity.
+
+## Use
+
+### Including the Middleware and setting Span names
+
+The middleware takes care of extracting the parent Span from the requests
+headers, both the [W3C](https://w3c.github.io/trace-context/) and [B3](https://github.com/openzipkin/b3-propagation) formats are supported.
+
+Because Elli has no router there is not a way to get a very descriptive Span
+name automatically. See the OpenTelemetry docs [Semantic conventions for HTTP spans](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name) for
+why you don't want to set the Span name to the raw path of the request. Thus,
+the Span has the name `"HTTP {METHOD_NAME}"`.
+
+The macro `update_name` from `opentelemetry_api/include/otel_tracer.hrl` allows you
+to update the name of the Span after it has started:
+
+``` erlang
+handle(Req, Args) ->
+    handle(Req#req.path, Req, Args).
+
+handle([<<"hello">>, Who], Req, _Args) ->
+    ?update_name(<<"/hello/{who}">>),
+    {ok, [], <<"Hello ", Who/binary>>}.
```

Comment:
This could possibly be done with a parse transform, but would require specifying a function/arity to always respect. Not required at this point for sure.

Reply:
Do you mean a pt that is able to take `[<<"hello">>, Who]` and create `<<"/hello/{Who}">>`? I like that. 

But what do you mean about specifying function/arity?

Reply:
yeah that's what I mean. The function arity would be that you go from handle/2 (which is probably registered or standard somewhere, I forget) to handle/3, so it's not guaranteed that handle/3 would be the callback that needs transforming.

Reply:
Ah, yea, if it were done I'd just define it as `handle/3`, requiring the user do the call from `handle/2` if they want to use it.